components/core/src/clp/hash_utils.hpp

```diff
@@ -0,0 +1,99 @@
+#ifndef CLP_HASH_UTILS_HPP
+#define CLP_HASH_UTILS_HPP
+
+#include <memory>
+#include <span>
+#include <string>
+#include <string_view>
+#include <vector>
+
+#include <openssl/evp.h>
+
+#include "ErrorCode.hpp"
+#include "TraceableException.hpp"
+
+namespace clp {
+/**
+ * A class the wraps around openssl EVP_MD_CTX to manage the life cycle of
+ * dynamically allocated EVP_MD_CTX object.
+ */
+class EvpCtxManager {
+public:
+    // Types
+    class OperationFailed : public TraceableException {
+    public:
+        // Constructors
+        OperationFailed(ErrorCode error_code, char const* const filename, int line_number)
+                : TraceableException(error_code, filename, line_number) {}
+
+        // Methods
+        [[nodiscard]] auto what() const noexcept -> char const* override {
+            return "EvpCtxManager operation failed";
+        }
+    };
+
+    // Constructors
+    EvpCtxManager() : m_md_ctx{EVP_MD_CTX_create()} {
+        if (nullptr == m_md_ctx) {
+            throw OperationFailed(ErrorCode_NotInit, __FILENAME__, __LINE__);
+        }
+    }
+
+    // Disable copy and move constructor/assignment
+    EvpCtxManager(EvpCtxManager const&) = delete;
+    auto operator=(EvpCtxManager const&) -> EvpCtxManager& = delete;
+
+    // Destructor
+    ~EvpCtxManager() { EVP_MD_CTX_destroy(m_md_ctx); }
+
+    auto digest_init_ex(const EVP_MD* type, ENGINE* impl) -> int {
+        return EVP_DigestInit_ex(m_md_ctx, type, impl);
+    }
+
+    auto digest_update(void const* d, size_t cnt) -> int {
+        return EVP_DigestUpdate(m_md_ctx, d, cnt);
+    }
+
+    auto digest_final_ex(unsigned char* md, unsigned int* s) -> int {
```

Comment:
> After calling EVP_DigestFinal_ex() no additional calls to EVP_DigestUpdate() can be made

According to the doc, we should have sth to prevent users from calling `update` after `final`. Might be worth adding a flag, and checking the flag in `update`.